sdk/metric/metrictest/exporter.go

```diff
@@ -0,0 +1,174 @@
+)
+
+// Exporter is a manually collected exporter for testing the SDK.  It does not
+// satisfy the `export.Exporter` interface because it is not intended to be
+// used with the periodic collection of the SDK, instead the test should
+// manually call `Collect()`
+//
+// Exporters are not thread safe, and should only be used for testing.
+type Exporter struct {
+	exports []ExportRecord
+	// resource *resource.Resource
+
+	controller *controller.Controller
+}
+
+func NewTestMeterProvider() (metric.MeterProvider, *Exporter) {
+	c := controller.New(
+		processor.NewFactory(
+			selector.NewWithHistogramDistribution(),
+			aggregation.CumulativeTemporalitySelector(),
+		),
+		controller.WithCollectPeriod(0),
+	)
+	exp := &Exporter{
+		controller: c,
+	}
+
+	return c, exp
+}
+
+// ExportRecord represents one collected datapoint from the Exporter.
+type ExportRecord struct {
+	InstrumentName         string
+	InstrumentationLibrary Library
+	Labels                 []attribute.KeyValue
+	AggregationKind        aggregation.Kind
+	Sum                    number.Number
+	Count                  uint64
+	Histogram              aggregation.Buckets
+	LastValue              number.Number
+}
+
+// Collect triggers the SDK's collect methods and then aggregates the data into
+// `ExportRecord`s.
+func (e *Exporter) Collect(ctx context.Context) error {
+	e.exports = []ExportRecord{}
+
+	err := e.controller.Collect(ctx)
+	if err != nil {
+		return err
+	}
+
+	return e.controller.ForEach(func(l instrumentation.Library, r export.Reader) error {
+		lib := Library{
+			InstrumentationName:    l.Name,
+			InstrumentationVersion: l.Version,
+			SchemaURL:              l.SchemaURL,
+		}
+
+		return r.ForEach(aggregation.CumulativeTemporalitySelector(), func(rec export.Record) error {
+			record := ExportRecord{
+				InstrumentName:         rec.Descriptor().Name(),
+				InstrumentationLibrary: lib,
+				Labels:                 rec.Labels().ToSlice(),
+				AggregationKind:        rec.Aggregation().Kind(),
+			}
+
+			var err error
+			switch agg := rec.Aggregation().(type) {
+			case aggregation.Histogram:
+				record.AggregationKind = aggregation.HistogramKind
+				record.Histogram, err = agg.Histogram()
+				if err != nil {
+					return err
+				}
+				record.Sum, err = agg.Sum()
+				if err != nil {
+					return err
+				}
+				record.Count, err = agg.Count()
+				if err != nil {
+					return err
+				}
+			case aggregation.Count:
+				record.Count, err = agg.Count()
+				if err != nil {
+					return err
+				}
+			case aggregation.LastValue:
+				record.LastValue, _, err = agg.LastValue()
+				if err != nil {
+					return err
+				}
+			case aggregation.Sum:
+				record.Sum, err = agg.Sum()
+				if err != nil {
+					return err
+				}
+			}
+
+			e.exports = append(e.exports, record)
+			return nil
+		})
+	})
+}
+
+// GetRecords returns all Records found by the SDK
+func (e *Exporter) GetRecords() []ExportRecord {
+	return e.exports
```

Comment:
Should this slice be copied before it is returned? If not should be just expose the field directly?